lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -54,225 +54,224 @@ typedef uint64_t uint64;
+            ICU_DEBUG_PRINT(ICU_ERROR_FMT, u_errorName(e));                   \
+            AssertMsg(false, u_errorName(e));                                 \
+            return r;                                                         \
+        }                                                                     \
+        else if (!(expr))                                                     \
+        {                                                                     \
+            ICU_DEBUG_PRINT(ICU_EXPR_FMT, u_errorName(e));                    \
+            Assert(expr);                                                     \
+            return r;                                                         \
+        }                                                                     \
+    } while (false)
+
+#define CHECK_UTF_CONVERSION(str, ret)                                        \
+    do                                                                        \
+    {                                                                         \
+        if (str == nullptr)                                                   \
+        {                                                                     \
+            AssertOrFailFastMsg(false, "OOM: failed to allocate string buffer"); \
+            return ret;                                                       \
+        }                                                                     \
+    } while (false)
+
 namespace PlatformAgnostic
 {
 namespace Intl
 {
     using namespace PlatformAgnostic::Resource;
 
-    bool IsWellFormedLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch)
+    // lots of logic below requires utf8char_t ~= char and UChar ~= char16
+    static_assert(sizeof(utf8char_t) == sizeof(char), "ICU-based Intl logic assumes that utf8char_t is compatible with char");
+
+    // [[ Ctrl-F: UChar_cast_explainer ]]
+    // UChar, like char16, is guaranteed to be 2 bytes on all platforms.
+    // However, we cannot use static_cast because on Windows:
+    // - char16 => WCHAR => wchar_t (native type) -- see Core/CommonTypedefs.h or Codex/Utf8Codex.h
+    // - (wchar_t is a native 2-byte type on Windows, but native 4-byte elsewhere)
+    // On other platforms:
+    // - char16 => char16_t -- see pal/inc/pal_mstypes.h
+    // All platforms:
+    // - UChar -> char16_t (all platforms)
+    static_assert(sizeof(UChar) == sizeof(char16), "ICU-based Intl logic assumes that UChar is compatible with char16");
+
+    // This function duplicates the Utf8Helper.h classes/methods
+    // See https://github.com/Microsoft/ChakraCore/pull/3913 for a (failed) attempt to use those directly
+    // Use CHECK_UTF_CONVERSION macro at the callsite to check the return value
+    // The caller is responsible for delete[]ing the memory returned here
+    static utf8char_t *Utf16ToUtf8(_In_ const char16 *src, _In_ const charcount_t srcLength, _Out_ size_t *destLength)
```

Comment:
Probably a better approach is to factor the bits that don't have unfortunate dependencies into a separate header and include that from Utf8Helper.h as well as here.

Reply:
(Rather than duplicating code which may become out of sync, leaving bugs in one of the versions.)

Reply:
Factor the bits of Utf8Helper with bad dependencies? I can look into it, but in the grand scheme of things, this code was concerning to me both because it was duplicated within this file, in addition to the fact that it duplicated the helper class somewhere else. I dont know how tricky it will be to get Utf8Helper to a state that it can be included here -- thoughts on logging it as its own issue (if it isnt already?) and coming back to it later?

Reply:
Not sure if we're saying the same thing, but this function works here, and it is logically a dependency of this code and of Utf8Helper.h (in the sense that that is where it lived before) -- so maybe you can factor out that dependency to another file.